contrib/generate-struct.py

```diff
@@ -0,0 +1,392 @@
+        for char in val:
+            if not char.isnumeric():
+                break
+            split_index += 1
+        if split_index:
+            self.multiplier = int(val[:split_index])
+
+        # find the type
+        try:
+            val2 = val[split_index:]
+            if val2.endswith("be"):
+                self.endian = Endian.BIG
+                val2 = val2[:-2]
+            elif val2.endswith("le"):
+                self.endian = Endian.LITTLE
+                val2 = val2[:-2]
+            self.type = Type(val2)
+        except ValueError as e:
+            raise ValueError(f"invalid type: {val2}") from e
+
+    def __str__(self) -> str:
+        tmp = f"{self.element_id}: "
+        if self.multiplier:
+            tmp += str(self.multiplier)
+        tmp += str(self.type)
+        if self.endian != Endian.NATIVE:
+            tmp += str(self.endian)
+        if self.default:
+            tmp += f": {self.default}"
+        elif self.constant:
+            tmp += f":: {self.constant}"
+        return tmp
+
+
+# convert a CamelCase name into snake_case
+def _camel_to_snake(name: str) -> str:
+
+    name_snake: str = ""
+    for char in name:
+        if char.islower():
+            name_snake += char
+            continue
+        if name_snake:
+            name_snake += "_"
+        name_snake += char.lower()
+    return name_snake
+
+
+class Generator:
+    def __init__(self, basename):
+        self.basename: str = basename
+
+    def _process_items(self, name: str, items: List[Item]) -> Tuple[str, str]:
+
+        # useful constants
+        name_snake = _camel_to_snake(name)
+
+        # header
+        str_h = "/* auto-generated, do not modify */\n"
+        str_h += "#include <glib.h>\n"
+        str_h += f"GByteArray* {name_snake}_new(void);\n"
+        str_h += f"GByteArray* {name_snake}_parse(const guint8 *buf, gsize bufsz, gsize offset, GError **error);\n"
+        for item in items:
+            if not item.enabled:
+                continue
+            # constants do not need getters and setters as they're, well, constant
+            if item.constant:
+                continue
+            str_h += f"{item.type_glib_mut} {name_snake}_get_{item.element_id}(GByteArray *st);\n"
+            if item.type == Type.STRING:
+                str_h += f"gboolean {name_snake}_set_{item.element_id}(GByteArray *st, {item.type_glib_const} value, GError **error);\n"
+            else:
+                str_h += f"void {name_snake}_set_{item.element_id}(GByteArray *st, {item.type_glib_const} value);\n"
+
+        # body
+        str_c = "/* auto-generated, do not modify */\n"
+        str_c += f'#include "{self.basename}"\n'
+        str_c += (
+            '#include "fu-byte-array.h"\n'  # FIXME: need override for libfwupdplugin.h
+        )
+        str_c += '#include "fu-mem.h"\n'  # FIXME: need override for libfwupdplugin.h
+        str_c += '#include "fu-string.h"\n'  # FIXME: need override for libfwupdplugin.h
```

Comment:
maybe the checker for the headers should ignore files with `struct.c` in the name?

Reply:
nah, there's a cute meson trick you can do to modify the local generator -- i'll do that in libfwupdplugin only -- atm I'm converting the tree like the FuStruct branch and finding out any missing parts.